shard-operator/src/main/java/com/redhat/service/bridge/shard/operator/watchers/ServiceEventSource.java

```diff
@@ -0,0 +1,82 @@
+package com.redhat.service.bridge.shard.operator.watchers;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.redhat.service.bridge.shard.operator.utils.LabelsBuilder;
+
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.WatcherException;
+import io.javaoperatorsdk.operator.processing.event.AbstractEventSource;
+import io.quarkus.runtime.Quarkus;
+
+import static io.javaoperatorsdk.operator.processing.KubernetesResourceUtils.getUID;
+import static io.javaoperatorsdk.operator.processing.KubernetesResourceUtils.getVersion;
+
+public class ServiceEventSource extends AbstractEventSource implements Watcher<Service> {
+    private static final Logger log = LoggerFactory.getLogger(ServiceEventSource.class);
+
+    private final KubernetesClient client;
+
+    private final String component;
+
+    public static ServiceEventSource createAndRegisterWatch(KubernetesClient client, String component) {
+        ServiceEventSource serviceEventSource = new ServiceEventSource(client, component);
+        serviceEventSource.registerWatch(component);
+        return serviceEventSource;
+    }
+
+    private ServiceEventSource(KubernetesClient client, String component) {
+        this.client = client;
+        this.component = component;
+    }
+
+    private void registerWatch(String component) {
+        client
+                .services()
+                .inAnyNamespace()
+                .withLabels(
+                        new LabelsBuilder()
+                                .withManagedByOperator()
+                                .withComponent(component)
+                                .build())
+                .watch(this);
+    }
+
+    @Override
+    public void eventReceived(Watcher.Action action, Service service) {
+        log.info(
+                "Event received for action: {}, Service: {}",
+                action.name(),
+                service.getMetadata().getName());
+
+        if (action == Watcher.Action.ERROR) {
+            log.warn(
+                    "Skipping {} event for custom resource uid: {}, version: {}",
+                    action,
+                    getUID(service),
+                    getVersion(service));
+            return;
+        }
+
+        eventHandler.handleEvent(new ServiceEvent(action, service, this));
+    }
+
+    @Override
+    public void onClose(WatcherException e) {
+        if (e == null) {
+            return;
```

Comment:
Is this happening in case of a graceful shutdown (aka without any error)?
What about adding a log?

Reply:
In theory this method is called when the watcher closes due to an Exception. The method `close()` is invoked when the watcher is gracefully closed instead. 

I'll add a log message

Reply:
Ok if this is the case, how it is possible that the watcher failed with an Exception but here there is no exception (aka `e == null`)?